multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ApplicationStopHandler.java

```diff
@@ -0,0 +1,43 @@
+package org.cloudfoundry.multiapps.controller.process.util;
+
+import java.time.Duration;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.cloudfoundry.multiapps.common.util.MiscUtil;
+import org.cloudfoundry.multiapps.controller.process.Messages;
+import org.cloudfoundry.multiapps.controller.process.steps.ProcessContext;
+
+import com.sap.cloudfoundry.client.facade.CloudControllerClient;
+import com.sap.cloudfoundry.client.facade.domain.CloudApplication;
+
+@Named
+public class ApplicationStopHandler {
+
+    protected ApplicationWaitAfterStopVariableGetter delayVariableGetter;
+
+    @Inject
+    public ApplicationStopHandler(ApplicationWaitAfterStopVariableGetter delayVariableGetter) {
+        this.delayVariableGetter = delayVariableGetter;
+    }
+
+    public void stopAppWithDelayIfConfigured(CloudApplication app, CloudControllerClient client, ProcessContext context) {
+        client.stopApplication(app.getName());
+        if (shouldWaitAfterAppStop(context)) {
+            Duration durationToWait = delayVariableGetter.getDelayDurationFromAppEnv(context);
+            context.getStepLogger()
+                   .info(Messages.WAITING_FOR_0_SECONDS, durationToWait.toSeconds());
+            MiscUtil.sleep(durationToWait.toMillis());
```

Comment:
One other approach for this is using flowable entities like [Timer](https://www.flowable.com/open-source/docs/bpmn/ch07b-BPMN-Constructs#timer-intermediate-catching-event) in bpmn diagram to wait some time before to continue with next flowable steps. In this way flowable step won't be in "busy" state and hold db transaction until Thread sleep finish.
At some point it's good that we have some restriction about maximum delay time because if we don't have such configuration, some client can make attack by applying very long stop delays and trigger multiple deployments which will consume all available resources.